### DISCUSSION TITLE
Question - Resolve the TypeName of a type argument with type arguments.

### DISCUSSION THREAD

**Discussion by callius** [square/kotlinpoet#1379] [Q&A] [answered]
The following is a class structure for example purposes:

```kotlin
interface Target<V> {
    val value: V 
}

interface TargetBuilder<I, T : Target<I>> {
    …
}

abstract class GenericTargetBuilder<I> : TargetBuilder<I, GenericTarget<I>> {
    …
}

class A : GenericTargetBuilder<Int>() {
    …
}
```

I would like to resolve `GenericTarget<Int>` from class `A` using kotlinpoet-ksp.

Following the [docs](https://square.github.io/kotlinpoet/interop-ksp/#type-parameters), I tried the following:
1. I do a tree search on the super types of class `A` to find the first `TypeReference` of `TargetBuilder`.
2. At this point, I have the `TypeReference`s and `TypeDeclaration`s of all the super types of class `A`.
3. I take the second type argument (index 1) of the `TargetBuilder` declaration on `GenericTargetBuilder`.
4. Now, I call `toTypeName` on the type argument type's `TypeReference`, passing in a recursive `TypeParametersResolver` (`foldRight` on all of the parent declarations’ type parameters, if any).

I get `GenericTarget<I>`, instead of the expected `GenericTarget<Int>`.

How should I go about resolving `GenericTarget<Int>` from this kind of class structure?

**Comment by callius** [comment, marked as answer]
I was able to solve my problem by calling `toClassName` and `parameterizedBy`, passing a list generated by recursively resolving each type parameter from the parent hierarchy by index.

Though, it would be nice to simply be able to use `toTypeName` for this.

**Reply by Egorand** [reply]
Are you interested in sending us a PR that adds the `toTypeName` method to cover this use case?

**Reply by callius** [reply]
@Egorand, possibly, although probably not any time soon.

However, in the time since this discussion was started, I published the library I had been working on which includes my aforementioned solution. If anyone comes across this thread and has some time on their hands, they are more than welcome to use it to create a PR.

My solution can be found [here](https://github.com/callius/target-kt/blob/c0b7a2e566bdf0705f6ee07305d9a42f18cdb518/target-libs/annotation-processor/src/main/kotlin/target/annotation_processor/ModelTemplateVisitorPoet.kt#L386).

Thanks!
